server/task.go

```diff
@@ -0,0 +1,277 @@
+	"github.com/arthurfabre/scheduler/server/pb"
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/clientv3util"
+	"github.com/golang/protobuf/proto"
+	"github.com/satori/go.uuid"
+	"log"
+	"strings"
+	"time"
+)
+
+// Format strings for prefixes. A prefix is a key with everything but the last Task UUID component
+// See README/#ETCD Key Schema
+const (
+	taskPrefix        = "task/"
+	queuedPrefixFmt   = "task/status/queued/"
+	runningPrefixFmt  = "task/status/running/%s/"
+	completePrefixFmt = "task/status/complete/%d/"
+	canceledPrefixFmt = "task/status/canceled/%d/"
+)
+
+// Task handles storing and updating tasks (and their status) in etcd.
+// Tasks embeds the pb.Task implementation generated by protoc.
+type Task struct {
+	*pb.Task
+
+	// Version in etcd when this was last updated / retrieved
+	version int64
+
+	// Key in etcd
+	key string
+}
+
+// watchQueuedTasks returns a Channel of Tasks that have just been queued
+// TODO - Potentially expose DELETE events, so we can backoff before stealing tasks
+func watchQueuedTasks(ctx context.Context, client *clientv3.Client) <-chan *Task {
+	out := make(chan *Task)
+
+	// TODO - Are we cleaning things up properly?
+	go func() {
+		for resp := range client.Watch(ctx, queuedPrefix(), clientv3.WithPrefix(), clientv3.WithFilterDelete()) {
+			for _, ev := range resp.Events {
+				task, err := getTask(ctx, client, taskID(string(ev.Kv.Key)))
+				if err != nil {
+					log.Println("Error retrieving task", ev.Kv.Key)
+					continue
+				}
+
+				out <- task
+			}
+		}
+	}()
+
+	return out
+}
+
+// listDoneTasks returns a list of Tasks that were done (completed or canceled) at least age seconds ago.
+func listDoneTasks(ctx context.Context, client *clientv3.Client, age int64) ([]*Task, error) {
+	// Get everything from epoch 0 to (Now - age)
+	end := time.Now().Unix() - age
+
+	completed, err := listTasks(ctx, client, completePrefix(0), clientv3.WithRange(completePrefix(end)))
+	if err != nil {
+		return nil, err
+	}
+
+	canceled, err := listTasks(ctx, client, canceledPrefix(0), clientv3.WithRange(canceledPrefix(end)))
+	if err != nil {
+		return nil, err
+	}
+
+	return append(completed, canceled...), nil
+}
+
+// listNodeTasks returns a list of Tasks that are being run by nodeId.
+func listNodeTasks(ctx context.Context, client *clientv3.Client, nodeId *api.NodeID) ([]*Task, error) {
+	return listTasks(ctx, client, runningPrefix(nodeId), clientv3.WithPrefix())
+}
+
+// listTasks returns a list of Tasks using etcd GET(key, opts...). Intended to be used with status keys.
+func listTasks(ctx context.Context, client *clientv3.Client, key string, opts ...clientv3.OpOption) ([]*Task, error) {
+	resp, err := client.Get(ctx, key, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	tasks := make([]*Task, 0, resp.Count)
+
+	for _, t := range resp.Kvs {
+		task, err := getTask(ctx, client, taskID(string(t.Key)))
+		if err != nil {
+			log.Println("Error retrieving task", t.Key)
```

Comment:
I think either the documentation should mention that tasks it failed to query are not included, or it should return both - partial task list and errors retrieving others (if any).